tests/fv/fv_test.go

```diff
@@ -85,52 +86,108 @@ var _ = Describe("PolicyController", func() {
 	})
 
-	Context("when etcd data is lost", func() {
-		BeforeEach(func() {
-			// Write some data
-			ns := &v1.Namespace{
-				ObjectMeta: metav1.ObjectMeta{
-					Name: "peanutbutter",
+	Context("policies", func() {
+		var policyName string
+		var genPolicyName string
+
+		// Create a Kubernetes NetworkPolicy.
+		policyName = "jelly"
+		genPolicyName = "knp.default.default." + policyName
+		var np *extensions.NetworkPolicy
+		np = &extensions.NetworkPolicy{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      policyName,
+				Namespace: "default",
+			},
+			Spec: extensions.NetworkPolicySpec{
+				PodSelector: metav1.LabelSelector{
+					MatchLabels: map[string]string{
+						"fools": "gold",
+					},
 				},
-				Spec: v1.NamespaceSpec{},
-			}
-			_, err := k8sClient.CoreV1().Namespaces().Create(ns)
-			Expect(err).NotTo(HaveOccurred())
-
-			// Wait for it to appear in etcd
-			Eventually(func() *api.Profile {
-				profile, _ := calicoClient.Profiles().Get(api.ProfileMetadata{Name: "k8s_ns.peanutbutter"})
-				return profile
-			}).ShouldNot(BeNil())
-		})
-		It("should recover by rewriting the data to etcd ", func() {
-			Stop(apiserver)
-			err := calicoClient.Profiles().Delete(api.ProfileMetadata{Name: "k8s_ns.peanutbutter"})
+			},
+		}
+		err := k8sClient.Extensions().RESTClient().
+			Post().
+			Resource("networkpolicies").
+			Namespace("default").
+			Body(np).
+			Do().Error()
+		Expect(err).NotTo(HaveOccurred())
+
+		// Wait for it to appear in Calico's etcd.
+		Eventually(func() *api.Policy {
+			policy, _ := calicoClient.Policies().Get(api.PolicyMetadata{Name: genPolicyName})
+			return policy
+		}).ShouldNot(BeNil())
+
+		It("should write new policies in etcd to match policies in k8s", func() {
+			// Delete the Policy.
+			err := calicoClient.Policies().Delete(api.PolicyMetadata{Name: genPolicyName})
 			Expect(err).ShouldNot(HaveOccurred())
 
-			Start(apiserver)
+			// Wait for the policy-controller to write it back.
 			Eventually(func() error {
-				_, err := calicoClient.Profiles().Get(api.ProfileMetadata{Name: "k8s_ns.peanutbutter"})
+				_, err := calicoClient.Policies().Get(api.PolicyMetadata{Name: genPolicyName})
 				return err
 			}, time.Second*15, 500*time.Millisecond).ShouldNot(HaveOccurred())
 		})
+
+		It("should update existing policies in etcd to match policies in k8s", func() {
+			// Change the selector of the policy in etcd.
+			_, err := calicoClient.Policies().Update(&api.Policy{
+				Metadata: api.PolicyMetadata{Name: genPolicyName},
+				Spec: api.PolicySpec{
+					Selector: "calico/k8s_ns == 'default' && ping == 'pong'",
+				},
+			})
+			Expect(err).ShouldNot(HaveOccurred())
+
+			// Wait for the policy-controller to set it back to its original value.
+			Eventually(func() string {
+				p, _ := calicoClient.Policies().Get(api.PolicyMetadata{Name: genPolicyName})
+				return p.Spec.Selector
+			}, time.Second*15, 500*time.Millisecond).Should(Equal("calico/k8s_ns == 'default' && fools == 'gold'"))
```

Comment:
@ozdanborne - can we also validate Types here? If you add the following to the NetworkPolicy, we should see `ingress` on the other end as well:
   `PolicyTypes: []extensions.PolicyType{extensions.PolicyTypeIngress},`

Reply:
I think this requires your changes though first, right @bcreane?

Reply:
Yes.

Reply:
Yes.